Add CI for running benchmarks

YJDoc2 opened the pull request:
Solves #324 

As discussed, I have added CI for running benchmarks each time a PR is opened, and subsequently, when anyone comments `/run-bench` on the PR. This will run the benchmarks and post the result table as a comment on the PR. On subsequent runs, the same comment will be edited and updated with new results.

~Because github needs the CI file to be present in default branch to run a CI, the bench CI won't run for this PR.~

The CI is failing here, because the comment post action needs write permission on PRs, which is not present for default github token provided to CI in this repo (?). I have opened a dummy PR [here](https://github.com/YJDoc2/comrak/pull/6) on my fork. you can see the posted results there, as well as try out the comment based invocation of the benchmarks.

I have also changed the benchmarking method as discussed. We used to simply clone the cmark-gfm and run the bench from there ; now I have copied the benchmark running script from there, and use hyperfine to run them. Now we also run benchmark on pulldown-cmark and cmark-gfm . Markdown-rs was also considered, but it is a library-only repo, so we will probably need to setup cargo project with simple binary which reads from stdin and runs the markdow-rs parse on it for running benchmarks on it. That is not done in this PR.

Some issues I have noticed :
- Benchmark CI takes ~5 minutes total to build and run benchmark on all repos. Major time is taken in building all of the binaries, the benchmarks only take ~1 minute total to run.
- The time measurement varies for each CI run considerably. Not sure if this is GH runner issue or what, because when running this locally, the measurements were consistent across several runs. What this would mean is that we cannot compare measurement from a previous CI run to current, but only the values in current CI run. This also means that the `relative` section in the measurement is most sensible for understanding difference and changes in speed. You can see the comment history of the measurement comment on my PR (linked above) to see what I mean by all this

YJDoc2 commented on benches/bench.sh:
```diff
@@ -0,0 +1,8 @@
+PROG=$1
+ROOTDIR=$(git rev-parse --show-toplevel)
+
+for lang in ar az be ca cs de en eo es es-ni fa fi fr hi hu id it ja ko mk nl no-nb pl pt-br ro ru sr th tr uk vi zh zh-tw; do \
+    cat $ROOTDIR/vendor/progit/$lang/*/*.markdown | $PROG > /dev/null
```
We don't use `seq 1..10` loop here like in original file from cmkar-gfm because instead the script is run several times by hyperfine itself

YJDoc2 commented on Makefile:
```diff
@@ -1,7 +1,43 @@
+ROOT:=$(shell git rev-parse --show-toplevel)
+COMMIT:=$(shell git rev-parse --short HEAD)
+MIN_RUNS:=25
```
To decide this value, I started with 10 runs,  and then checked 50 runs. the difference in measurements were not that much, even the std dev hadn't changed that much from 10 runs to 50. So I settled on 25.

YJDoc2 commented on .github/workflows/benchmarks.yml:
```diff
@@ -0,0 +1,36 @@
+    permissions:
+      pull-requests: write
+    # run  either when pull request is opened or when comment body (only on pr) is /run-bench
+    if: (github.event_name == 'pull_request') || ((github.event.issue.pull_request != null) && github.event.comment.body == '/run-bench')
```
I can change the comment body from `/run-bench` to something else if you want.

YJDoc2 commented:
@kivikakk @digitalmoksha Please take a look when possible :)

digitalmoksha commented on Makefile:
```diff
@@ -1,7 +1,43 @@
+	cargo build --release && \
+	cp target/release/pulldown-cmark ${ROOT}/benches/pulldown-cmark
+
+bench-comrak: build-comrak-branch
```
wdyt about adding [markdown-it.rs](https://github.com/rlidwka/markdown-it.rs) to the mix? It would be nice to have [markdown-rs](https://github.com/wooorm/markdown-rs), but as you said, at the moment there is not a command line version.

One thing we have to remember is that this really isn't testing exactly apples to apples. For example `markdown-rs` doesn't support footnotes and task lists directly yet. Though some [plugins](https://github.com/chrisjsewell/markdown-it-plugins.rs) have just been added for that.

A follow-on PR might have a small harness program for the Rust implementations, which would link in all the different `lib` versions.  The harness would take a switch, and then run the specified implementation, with the proper switches set to make them all use similar functionality.

YJDoc2 replied:
Sure, I didn't know about it ; so will take a look and try adding it as well.

digitalmoksha reviewed (commented):
@YJDoc2 this looks pretty good from my angle! Just had a few comments.

kivikakk commented:
This is looking amazing! Thank you so much.

I echo @digitalmoksha's comments about the use of `sudo` — is there usually a measurable difference between clearing the caches for our purposes? I imagine the warmup runs will remove a lot of the startup variation even without doing so.

> * The time measurement varies for each CI run considerably. Not sure if this is GH runner issue or what, because when running this locally, the measurements were consistent across several runs. What this would mean is that we cannot compare measurement from a previous CI run to current, but only the values in current CI run. This also means that the `relative` section in the measurement is most sensible for understanding difference and changes in speed. You can see the comment history of the measurement comment on my PR (linked above) to see what I mean by all this

Yes, that makes a lot of sense. I'm hopeful the relative data itself is accurate enough, often enough — the inter-run differences are probably caused by being scheduled onto hosts with vastly different loads at that time, but there's no predicting when that might happen during a run.

YJDoc2 commented on .github/workflows/benchmarks.yml:
```diff
@@ -0,0 +1,36 @@
+      - name: Post result comment
+        uses: mshick/add-pr-comment@v2
+        with:
+          message-path: bench-output.md
```
Hey, good idea! I'll add a line in the comment showing when the CI was run date+time (GMT). Thanks for the suggestion :+1:

YJDoc2 commented:
Hey @kivikakk @digitalmoksha , thanks for the review :)

>Hmm, do we really need to use sudo? I tried make bench-comrak locally, and hyperfine just hung during warmup, and no idea why. Turns out sudo was asking for a password. I don't really feel comfortable giving a bench mark program sudo access.

>I echo @digitalmoksha's comments about the use of sudo — is there usually a measurable difference between clearing the caches for our purposes? I imagine the warmup runs will remove a lot of the startup variation even without doing so.

I added the cache clearing step because otherwise hyperfine gave a warning that "first run is an outlier, which might be an effect of cache". Also when I tried without cache clearing, the measurements were considerably different. (consistent, but different). That said, I don't like the sudo use either, especially for benchmarks, so I'll try to figure out a workaround. Maybe increasing the number of warmup runs might solves the issue. Will get back after updating this.

Thanks :)

YJDoc2 commented:
@kivikakk @digitalmoksha I have removed sudo and added markdown-it rs in comparisons. The benchmark value change from using the cache drops, but it is consistent within itself, so it should be fine.  I have also added timestamp in the comment. Please take a look.

I'll see why the CI is failing...

digitalmoksha commented on README.md:
```diff
@@ -197,6 +197,33 @@ assert_eq!(
+
+This will build comrak in release mode, and run benchmark on it. You will see the time measurements as reported by hyperfine in the console.
+
+Makefile also provides a way to run benchmarks for comark current state (with your changes), comrak main branch, cmark-gfm and pulldown-cmark. For this you will need to install cmake. After that make sure that you have set-up the git submodules. In case you have not installed submodules when cloning, you can do it by running
```
Did you want to add a mention of `markdown-it.rs` here?

I added links to the repos as well, hope you don't mind.

```suggestion
Makefile also provides a way to run benchmarks for `comrak` current state (with your changes), `comrak` main branch, [`cmark-gfm`](https://github.com/github/cmark-gfm), [`pulldown-cmark`](https://github.com/raphlinus/pulldown-cmark) and [`markdown-it.rs`](https://github.com/rlidwka/markdown-it.rs). For this you will need to install `cmake`. After that make sure that you have set-up the git submodules. In case you have not installed submodules when cloning, you can do it by running
```

digitalmoksha commented on README.md:
```diff
@@ -197,6 +197,33 @@ assert_eq!(
+
+which will run benchmarks across all, and report the time take by each as well as relative time.
+
+Apart from this, CI is also setup for running benchmarks when a pull request is first opened. It will comment the results on the pull request in a tabular format comparing the 4 versions. After that you can manually trigger this CI by commenting `/run-bench` on the PR, this will update the existing comment with new results. Note Benchmarks won't be automatically run on each push.
```
Just a suggestion

```suggestion
Apart from this, CI is also setup for running benchmarks when a pull request is first opened. It will add a comment with the results on the pull request in a tabular format comparing the 5 versions. After that you can manually trigger this CI by commenting `/run-bench` on the PR, this will update the existing comment with new results. Note benchmarks won't be automatically run on each push.
```

digitalmoksha commented on README.md:
```diff
@@ -197,6 +197,33 @@ assert_eq!(
+
+For running benchmarks, you will need to [install hyperfine](https://github.com/sharkdp/hyperfine#installation) and optionally cmake.
+
+If you want to just run benchmark for comrak, with current state of repo, you can simply run
```
Minor suggestion

```suggestion
If you want to just run the benchmark for `comrak`, with the current state of the repo, you can simply run
```

digitalmoksha commented:
@YJDoc2 this look good to me, nice job.

kivikakk commented:
I'll look into the CI failure this afternoon with a view to merging this.

YJDoc2 commented:
Hey, sorry I wasn't able to update on this last week.

> I'll look into the CI failure this afternoon with a view to merging this.

Thanks a lot. Is it ok if I rebase it with main and force push? Maybe the CI is expecting some new things in master that aren't pulled in this branch.

YJDoc2 commented:
@kivikakk I have updated the proc-macro2 version, as that was causing the build failure on nightly.
Related issue https://github.com/dtolnay/proc-macro2/issues/356 .
I haven't updated anything else in cargo lock.

kivikakk reviewed (approved):
Looks excellent. Tested all locally — thank you so much for your effort! 🤍
